charmcraft/commands/store/__init__.py

```diff
@@ -113,3 +116,69 @@ def run(self, parsed_args):
         table = tabulate(data, headers=headers, tablefmt='plain')
         for line in table.splitlines():
             logger.info(line)
+
+
+class UploadCommand(BaseCommand):
+    """Upload a charm file to the Store."""
+    name = 'upload'
+    help_msg = "upload a charm file to the Store"
+
+    def _discover_charm(self, charm_filepath):
+        """Discover the charm name and file path.
+
+        If received path is None, a metadata.yaml will be searched in the current directory. If
+        path is given the name is taken from the filename.
+
+        """
+        if charm_filepath is None:
+            # discover the info using project's metadata, asume the file has the project's name
+            # with a .charm extension
+            try:
+                with open('metadata.yaml', 'rb') as fh:
+                    metadata = yaml.safe_load(fh)
+                charm_name = metadata['name']
+            except (yaml.error.YAMLError, OSError, KeyError):
+                raise CommandError(
+                    "Can't access name in 'metadata.yaml' file. The 'upload' command needs to be "
+                    "executed in a valid project's directory, or point to a charm file with "
+                    "the --charm-file option.")
+
+            charm_filepath = pathlib.Path(charm_name + '.charm').absolute()
+            if not os.access(charm_filepath, os.R_OK):  # access does not support pathlib in 3.5
```

Comment:
you added the comment, but you didn't `str()` the `Path` :-)

Reply:
You could also do `charm_filepath.exists()` tho

Reply:
That it exists doesn't mean that you can read it.

Fixed it.